File: src/controllers/slackController.ts
import { respondToUser } from "./../services/openAIService";
import {
  sendMessage,
  getThreadMessages,
  getThreadMessagesWithUsernames,
} from "../services/slackService";
import { getConversationSummary } from "../services/openAIService";
import { Request, Response } from "express";
import { AppMentionPayload } from "seratch-slack-types/events-api";

export const postEvent = async (req: Request, res: Response) => {
  if (req.body.type === "url_verification") {
    res.send(req.body.challenge);
    return;
  }

  const payload = req.body as AppMentionPayload;
  // console.log(payload);

  const { channel, text, thread_ts, ts } = payload.event;
  const botId = findBotId(payload);
  if (!botId) {
    console.error("botId not found");
    res.sendStatus(500);
    return;
  }

  try {
    res.sendStatus(200);
    let threadToReply = thread_ts;
    // Reply in a new thread if the message is not in a thread
    if (thread_ts !== ts) {
      threadToReply = ts;
    }
    const userMessage = text.replace(`<@${botId}>`, "").trim();
    let messages = [`USER:${userMessage}`];
    if (thread_ts) {
      messages =
        (await getThreadMessagesWithUsernames(channel, thread_ts, botId)) ||
        messages;
    }
    console.log({ messages });
    const response = await respondToUser(messages);

    await sendMessage(channel, threadToReply, response);
    return;
  } catch (error) {
    console.error(error);
  }
};

export const postCommand = async (req: Request, res: Response) => {
  const command = req.params.commandName;

  try {
    switch (command) {
      case "test": {
        res.status(200).send("This is a test command");
      }
      default: {
        const error = `Sorry, I don't know how to handle the command '${command}'`;
        res.status(200).send(error);
      }
    }
  } catch (error) {
    console.error(error);
    res
      .status(500)
      .send({ error: "An error occured while processing the request" });
  }
};

const getThreadSummary = async (channelId, threadTs) => {
  const threadMessages = await getThreadMessages(channelId, threadTs);
  const summary = await getConversationSummary(threadMessages);

  return summary;
};

const findBotId = (payload: AppMentionPayload) => {
  const bot = payload.authorizations.find((auth) => auth.is_bot);
  return bot.user_id;
};

File: py-src/controllers/slack_controller.py
from flask import jsonify
import time
from services.slack_service import (
    send_message,
    get_thread_messages,
    get_thread_messages_with_usernames,
)
from services.openai_service import respond_to_user
from threading import Thread

# POST /slack/events
def post_event(request):
    if request.get("type") == "url_verification":
        return request.get("challenge")

    Thread(target=process_event_payload, args=(request,)).start()
    return jsonify({"message": "success"}), 200


def find_bot_id(payload):
    for auth in payload["authorizations"]:
        # Check if the current authorization is a bot
        if auth["is_bot"]:
            return auth["user_id"]

    return None


def process_event_payload(payload):
    event = payload.get("event")
    channel = event.get("channel")
    text = event.get("text")
    thread_ts = event.get("thread_ts")
    ts = event.get("ts")
    bot_id = find_bot_id(payload)
    if not bot_id:
        print("botId not found")
        return
    try:
        thread_to_reply = thread_ts
        if thread_ts != ts:
            thread_to_reply = ts
        messages = f"USER: {text.replace(f'<@{bot_id}>','').strip()}"
        if thread_ts:
            messages = (
                get_thread_messages_with_usernames(channel, thread_ts, bot_id)
                or messages
            )

        start_time = time.perf_counter()
        response = respond_to_user(messages)
        end_time = time.perf_counter()
        response = response + f"\n\n\n_response generated in {round(end_time - start_time, 2)}s_ :hourglass:"

        return send_message(channel, thread_to_reply, response)
    except Exception as error:
        # Improve error handling
        print(error)
        return
